### ISSUE TITLE
[Research] Contrastive Learning for Cross-Subject EEG Emotion Transfer (EmotionCLIP / CSCL)

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#40]
## Paper 1
**Title:** Cross-subject EEG signals-based emotion recognition using contrastive learning
**Authors:** Various
**Year:** 2025
**Venue:** Scientific Reports (Nature)
**URL:** https://www.nature.com/articles/s41598-025-13289-5

## Paper 2
**Title:** Cross-domain EEG-based Emotion Recognition with Contrastive Learning
**Year:** 2025
**Venue:** arXiv (2511.05293)
**URL:** https://arxiv.org/abs/2511.05293

## Key Contribution
**CSCL (Cross-Subject Contrastive Learning)** learns emotion-discriminative EEG representations in hyperbolic space using emotion and stimulus contrastive losses. Aligns EEG signals from different subjects across time, space, and frequency domains, mitigating cross-subject distribution shift.

**CL-CS** generates positive/negative pairs by computing correlations between EEG signals — same emotional stimulus = positive pair, different = negative. Maximizes within-emotion correlation while minimizing cross-emotion correlation.

**EmotionCLIP** reformulates emotion recognition as EEG-text matching (CLIP framework). Achieves cross-subject accuracies of **88.69% on SEED** and **73.50% on SEED-IV** — state-of-the-art for cross-subject.

## Relevance to NeuralDreamWorkshop
The #1 accuracy bottleneck is the cross-subject gap (within-subject 90.97% vs cross-subject 64.82% = 26-point drop). Contrastive pre-training learns subject-invariant EEG representations that dramatically narrow this gap. This directly addresses the "No personalization" known issue.

Current mega_lgbm model at 74.21% CV could be replaced/augmented with contrastively pre-trained features → projected 80-88% cross-subject.

### Files to Create
- `ml/models/contrastive_eeg.py` — Contrastive pre-training framework:
  ```python
  class EEGContrastiveEncoder(nn.Module):
      """Pre-trains subject-invariant EEG representations."""
      def __init__(self, n_channels=4, n_features=41, proj_dim=128):
          # Feature encoder: 1D-CNN or Transformer
          # Projection head: MLP → L2-normalized embeddings
          # Contrastive loss: NT-Xent (normalized temperature-scaled)

  class ContrastivePreTrainer:
      """Train encoder using emotion/stimulus contrastive losses."""
      def train(self, datasets: List[EEGDataset]):
          # Same emotion from different subjects → positive pairs
          # Different emotions → negative pairs
          # Align representations across subjects
  ```
- `ml/training/train_contrastive.py` — Pre-training script on all available datasets (DEAP, SEED, GAMEEMO, DREAMER)
- `ml/training/finetune_contrastive.py` — Fine-tune pre-trained encoder for Muse 2 with linear probe

### Files to Modify
- `ml/models/emotion_classifier.py`:
  - Add contrastive encoder as new inference path (priority above mega_lgbm when available):
    `contrastive → mega_lgbm → ONNX → sklearn → feature-based`
  - Load pre-trained encoder + linear classifier head
- `ml/training/mega_trainer.py` — Add contrastive pre-training stage before supervised training
- `ml/api/routes.py` — Report `model_type: "contrastive"` in emotion responses
- `ml/processing/eeg_processor.py` — Add `extract_de_features_multichannel()` for differential entropy features (preferred input for contrastive models)

### Training Pipeline
```
Stage 1: Contrastive Pre-Training (unsupervised)
  - Input: All EEG datasets (DEAP + SEED + GAMEEMO + DREAMER)
  - Loss: NT-Xent with emotion-aware positive pairs
  - Output: Pre-trained EEG encoder (subject-invariant)
  - Duration: ~2-4 hours on GPU

Stage 2: Linear Probing (supervised)
  - Freeze encoder, train linear classifier head
  - Train on labeled emotion datasets
  - Output: emotion_contrastive_model.pt

Stage 3: Per-User Fine-Tuning (optional)
  - Unfreeze last layer of encoder
  - Fine-tune on 5-min user-specific data
  - Reduces cross-subject gap by another 5-10%
```

## Expected Impact
- **Cross-subject accuracy:** 80-88% (up from 74.21% — 6-14 point improvement)
- **Per-user fine-tuned:** 85-92% (approaching within-subject performance)
- **Key win:** Subject-invariant representations transfer across users without personalization
- **Bonus:** Pre-trained encoder improves ALL downstream tasks (sleep, attention, stress)

## Complexity: HIGH
- Requires GPU for contrastive pre-training (2-4 hours)
- Implementation: ~500 lines for encoder + training loop + loss functions
- Multi-dataset loading and alignment
- Careful hyperparameter tuning (temperature, learning rate, batch size)
